### ISSUE TITLE
Emit ordinal edges for upper bounds of Java type params. [phab:D1617]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2041]
_Created by **benjyw** at 2017-05-04 19:13:29_: 

The order matters - it affects type erasure.
Addresses #2035.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-05-18 17:58:37_:

Ping on this? Thanks.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-05-19 14:20:35_:

Seems reasonable to me. zarko to review implications on the schema.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-05-19 20:26:51_:

Does the order matter for upper bounds that *aren't* the first bound? If not, what do we gain by recording the distinction between them?

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-05-19 21:49:04_:

I believe that in Java the order (other than the first bound) doesn't matter.

But it does seem to matter in other languages, e.g., in Scala:

http://www.scala-lang.org/files/archive/spec/2.11/03-types.html#equivalence

```
Two compound types are equivalent if the sequences of their component are pairwise equivalent, and occur in the same order.
```

So it seems like using ordinals is schematically sound, rather than introducing some ad-hoc notation for Java, and still having to add ordinals later for Scala. It also just seems useful to me (for debugging, user-facing visualization and other purposes) to be able to recover the source order from the Kythe graph. 

Does that make sense?

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-06-05 21:40:02_:

Ping on this? Thanks.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-06-05 23:03:14_

**Comment by kythe** [comment]
_**zarko** closed via commit 66edbb53f868483f64629b1864a37ab85b6f2f0c at 2017-06-05 23:04:11_
